### PR TITLE
[ie/Turbo] Fix turbo extractor

### PR DESCRIPTION
**Pull request by nbr23**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

The Turbo site has changes, and the extractor is currently broken. This fixes it, and adds support for "playlists" (through their RSS feeds).

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by seproDev** [comment]
I don't think the extractor should be changed like this to add playlist support. If you do want to add playlist support, that should be done in a dedicated extractor. However, I am not convinced that adding the RSS feeds like this is super useful, as most posts don't contain videos. In its current form, the code also seems broken? Do you have an example where parsing the RSS feeds would be useful?

Looking at the site, it may be worth to split viously.com in to its own embed extractor instead. That would also deal more easily with the case where a single page contains multiple videos. And as a added bonus, it would also work on other sites using that CDN.

**Comment by nbr23** [comment]
Thanks a bunch for the feedback @seproDev, it all makes perfect sense to me!
I've made this into a viously.com embed extractor (this actually solves the RSS feed thing by design, so I love it).

Removing the old Turbo extractor too since it's broken anyway.

File: yt_dlp/extractor/viously.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            'upload_date': str,
+            'timestamp': float,
+        }
+    }]
+
+    def _extract_from_webpage(self, url, webpage):
+        has_vously_player = get_element_html_by_class('viously-player', webpage) or get_element_html_by_class('vsly-player', webpage)
+        if not has_vously_player:
+            return
+        viously_players = get_elements_html_by_class('viously-player', webpage) + get_elements_html_by_class('vsly-player', webpage)
```
Comment (seproDev) [inline yt_dlp/extractor/viously.py:29]: I think we can just
```suggestion
        viously_players = (get_elements_html_by_class('viously-player', webpage)
                           + get_elements_html_by_class('vsly-player', webpage))
        if not viously_players:
            return

```

File: yt_dlp/extractor/viously.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        for viously_player in viously_players:
+            video_id = self._html_search_regex(r'id="([-_\w]+)"', viously_player, 'video_id')
+            title = self._html_extract_title(webpage)
+            yield {
+                'id': video_id,
+                'title': title,
+                'description': title,
+                'formats': self._extract_m3u8_formats(self._API_URL.format(video_id), video_id),
+            }
```
Comment (seproDev) [inline yt_dlp/extractor/viously.py:38]: Here I would actually call their API to extract the metadata
```suggestion
        def custom_decode(text):
            STANDARD_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/='
            CUSTOM_ALPHABET = 'VIOUSLYABCDEFGHJKMNPQRTWXZviouslyabcdefghjkmnpqrtwxz9876543210+/='
            data = base64.b64decode(text.translate(str.maketrans(CUSTOM_ALPHABET, STANDARD_ALPHABET)))
            return data.decode('utf-8').strip('\x00')

        for video_id in traverse_obj(viously_players, (..., {extract_attributes}, 'id')):
            data = self._download_json(
                f'https://www.viously.com/export/json/{video_id}', video_id, transform_source=custom_decode)
            yield {
                'id': video_id,
                'formats': self._extract_m3u8_formats(f'https://www.viously.com/video/hls/{video_id}/index.m3u8', video_id),
                **traverse_obj(data, ('video', {
                    'title': 'title',
                    'description': 'description',
                    'duration': ('duration', {int_or_none}),
                    'timestamp': ('iso_date', {parse_iso8601}),
                    'categories': ('category', {lambda x: [x['name']]}),
                })),
            }
```
Tests will need to be updated.

File: yt_dlp/extractor/viously.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+from .common import InfoExtractor
+from ..utils import (
+    get_element_html_by_class,
+    get_elements_html_by_class,
+)
```
Comment (seproDev) [inline yt_dlp/extractor/viously.py:5]:
```suggestion
import base64

from .common import InfoExtractor
from ..utils import (
    extract_attributes,
    get_elements_html_by_class,
    int_or_none,
    parse_iso8601,
)
from ..utils.traversal import traverse_obj
```

**Review by seproDev — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/viously.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+
+class ViouslyIE(InfoExtractor):
+    _VALID_URL = False
+    _API_URL = 'https://www.viously.com/video/hls/{0:}/index.m3u8'
```
Comment (seproDev) [inline yt_dlp/extractor/viously.py:10]:
```suggestion
```

File: yt_dlp/extractor/viously.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            'title': 'Turbo du 07/09/2014 : Renault Twingo 3, Bentley Continental GT Speed, CES, Guide Achat Dacia...',
+            'description': 'Turbo du 07/09/2014 : Renault Twingo 3, Bentley Continental GT Speed, CES, Guide Achat Dacia...',
+            'age_limit': 0,
+            'upload_date': str,
+            'timestamp': float,
+        }
+    }]
+
+    def _extract_from_webpage(self, url, webpage):
+        viously_players = get_elements_html_by_class('viously-player', webpage) + get_elements_html_by_class('vsly-player', webpage)
```
Comment (seproDev) [inline yt_dlp/extractor/viously.py:30]: Sorry, this should use `re.findall`. The `get_element_by ` functions can throw on malformed HTML, so they could throw on any generic webpage.
```suggestion
        viously_players = re.findall(r'<div[^>]*class="(?:[^"]*\s)?v(?:iou)?sly-player(?:\s[^"]*)?"[^>]*>', webpage)
```

File: yt_dlp/extractor/viously.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        'url': 'http://www.turbo.fr/videos-voiture/454443-turbo-du-07-09-2014-renault-twingo-3-bentley-continental-gt-speed-ces-guide-achat-dacia.html',
+        'md5': '37a6c3381599381ff53a7e1e0575c0bc',
+        'info_dict': {
+            'id': 'F_xQzS2jwb3',
+            'ext': 'mp4',
+            'title': 'Turbo du 07/09/2014 : Renault Twingo 3, Bentley Continental GT Speed, CES, Guide Achat Dacia...',
+            'description': 'Turbo du 07/09/2014 : Renault Twingo 3, Bentley Continental GT Speed, CES, Guide Achat Dacia...',
+            'age_limit': 0,
+            'upload_date': str,
+            'timestamp': float,
```
Comment (seproDev) [inline yt_dlp/extractor/viously.py:25]:
```suggestion
            'title': 'Turbo du 07/09/2014\xa0: Renault Twingo 3, Bentley Continental GT Speed, CES, Guide Achat Dacia...',
            'description': 'Turbo du 07/09/2014\xa0: Renault Twingo 3, Bentley Continental GT Speed, CES, Guide Achat Dacia...',
            'age_limit': 0,
            'upload_date': '20230328',
            'timestamp': 1680037507,
            'duration': 3716,
            'categories': ['motors'],
```

**Review by seproDev — CHANGES_REQUESTED** [review]
Please don't force push

File: yt_dlp/extractor/viously.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+            return data.decode('utf-8').strip('\x00')
+
+        for video_id in traverse_obj(viously_players, (..., {extract_attributes}, 'id')):
+            formats = self._extract_m3u8_formats(
+                f'https://www.viously.com/video/hls/{video_id}/index.m3u8', video_id, fatal=False)
+            data = self._download_json(
+                f'https://www.viously.com/export/json/{video_id}', video_id,
+                transform_source=custom_decode, fatal=False)
+            if not formats or not data:
+                continue
```
Comment (bashonly) [inline yt_dlp/extractor/viously.py:49]: we can still extract if we have formats w/o metadata
```suggestion
            formats = self._extract_m3u8_formats(
                f'https://www.viously.com/video/hls/{video_id}/index.m3u8', video_id, fatal=False)
            if not formats:
                continue
            data = self._download_json(
                f'https://www.viously.com/export/json/{video_id}', video_id,
                transform_source=custom_decode, fatal=False)
```

File: yt_dlp/extractor/viously.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            yield {
+                'id': video_id,
+                'formats': formats,
+                **traverse_obj(data, ('video', {
+                    'title': 'title',
+                    'description': 'description',
+                    'duration': ('duration', {int_or_none}),
+                    'timestamp': ('iso_date', {parse_iso8601}),
+                    'categories': ('category', {lambda x: [x['name']]}),
```
Comment (bashonly) [inline yt_dlp/extractor/viously.py:58]:
```suggestion
                    'categories': ('category', 'name', {str}, {lambda x: [x] if x else None}),
```
or else we could get `[None]` and `['']` etc

File: yt_dlp/extractor/viously.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                f'https://www.viously.com/export/json/{video_id}', video_id,
+                transform_source=custom_decode, fatal=False)
+            if not formats or not data:
+                continue
+            yield {
+                'id': video_id,
+                'formats': formats,
+                **traverse_obj(data, ('video', {
+                    'title': 'title',
+                    'description': 'description',
```
Comment (bashonly) [inline yt_dlp/extractor/viously.py:55]:
```suggestion
                    'title': ('title', {str}),
                    'description': ('description', {str}),
```

**Review by bashonly — CHANGES_REQUESTED** [review]

**Review by bashonly — APPROVED** [review]

**Review by seproDev — APPROVED** [review]

**Pull request merged**